File: code/pipeline-cpp/src/os/allocator.cpp
#include <cstdlib>
#include <new>

#include "allocator.h"

/// Global atomic counters to track allocated and freed bytes, as well as
/// allocation count.
namespace telemetry {
std::atomic<size_t> allocated_bytes{ 0 };
std::atomic<size_t> allocation_count{ 0 };
std::atomic<size_t> freed_bytes{ 0 };

thread_local bool track_allocations = true;
}

/// \brief Global operator new overload to track memory allocations.
///
/// \param count The number of bytes to allocate.
/// \return A pointer to the allocated memory.
void*
operator new(std::size_t count)
{
  if (telemetry::track_allocations) {
    telemetry::allocated_bytes.fetch_add(count, std::memory_order_relaxed);
    telemetry::allocation_count.fetch_add(1, std::memory_order_relaxed);
  }

  if (void* ptr = std::malloc(count)) {
    return ptr;
  }

  throw std::bad_alloc{};
}

/// \brief Global operator new overload to track memory allocations.
///
/// \param count The number of bytes to allocate.
/// \return A pointer to the allocated memory.
void*
operator new[](std::size_t count)
{
  return operator new(count);
}

/// \brief Global operator delete overload to track memory deallocations.
///
/// No exception variation.
///
/// \param ptr Pointer to the memory to deallocate.
/// \param count The number of bytes to deallocate.
void
operator delete(void* ptr, std::size_t count) noexcept
{
  if (telemetry::track_allocations) {
    telemetry::freed_bytes.fetch_add(count, std::memory_order_relaxed);
  }

  std::free(ptr);
}

/// \brief Global operator delete overload to track memory deallocations.
///
/// \param ptr Pointer to the memory to deallocate.
/// \param count The number of bytes to deallocate.
void
operator delete[](void* ptr, std::size_t count) noexcept
{
  operator delete(ptr, count);
}

/// \brief Global operator delete overload.
///
/// Required for safety.
///
/// \param ptr Pointer to the memory to deallocate.
void
operator delete(void* ptr) noexcept
{
  std::free(ptr);
}

/// \brief Global operator delete overload.
///
/// Required for safety.
///
/// \param ptr Pointer to the memory to deallocate.
void
operator delete[](void* ptr) noexcept
{
  std::free(ptr);
}

/// \brief Global operator new overload to track memory allocations with
/// alignment.
///
/// Required for C++17 aligned allocation.
///
/// \param count The number of bytes to allocate.
/// \param al The alignment requirement for the allocation.
/// \return A pointer to the allocated memory.
void*
operator new(std::size_t count, std::align_val_t al)
{
  if (telemetry::track_allocations) {
    telemetry::allocated_bytes.fetch_add(count, std::memory_order_relaxed);
    telemetry::allocation_count.fetch_add(1, std::memory_order_relaxed);
  }

  // Linux fix: round up 'count' to the nearest multiple of 'alignment'
  std::size_t alignment = static_cast<std::size_t>(al);
  std::size_t fit_size = (count + alignment - 1) & ~(alignment - 1);

  if (void* ptr = ::aligned_alloc(alignment, fit_size)) {
    return ptr;
  }

  throw std::bad_alloc{};
}

/// \brief Global operator new overload to track array memory allocations with
/// alignment.
///
/// \param count The number of bytes to allocate.
/// \param al The alignment requirement for the allocation.
/// \return A pointer to the allocated memory.
void*
operator new[](std::size_t count, std::align_val_t al)
{
  return operator new(count, al);
}

/// \brief Global operator delete overload to track memory deallocations with
/// alignment.
///
/// Required for C++17 aligned deallocation.
///
/// \param ptr Pointer to the memory to deallocate.
/// \param count The number of bytes to deallocate.
/// \param al The alignment requirement for the deallocation.
void
operator delete(void* ptr,
  std::size_t count,
  [[maybe_unused]] std::align_val_t al) noexcept
{
  if (telemetry::track_allocations) {
    telemetry::freed_bytes.fetch_add(count, std::memory_order_relaxed);
  }

  std::free(ptr);
}

/// \brief Global operator delete overload to track array memory deallocations
/// with alignment.
///
/// \param ptr Pointer to the memory to deallocate.
/// \param count The number of bytes to deallocate.
/// \param al The alignment requirement for the deallocation.
void
operator delete[](void* ptr, std::size_t count, std::align_val_t al) noexcept
{
  operator delete(ptr, count, al);
}

File: code/pipeline-rust/src/os/allocator.rs
use std::alloc::{GlobalAlloc, Layout, System};
use std::cell::Cell;
use std::sync::atomic::{AtomicUsize, Ordering};

/// A custom global allocator that tracks memory allocations and deallocations.
pub struct TrackingAllocator;

/// Global atomic counters to track allocated and freed bytes, as well as
/// allocation count.
pub static ALLOCATED_BYTES: AtomicUsize = AtomicUsize::new(0);
pub static FREED_BYTES: AtomicUsize = AtomicUsize::new(0);
pub static ALLOCATION_COUNT: AtomicUsize = AtomicUsize::new(0);

thread_local! {
    pub static TRACK_ALLOCATIONS: Cell<bool> = Cell::new(true);
}

pub struct ScopedToggle {
    previous_state: bool,
}

impl ScopedToggle {
    pub fn new() -> Self {
        let previous_state = TRACK_ALLOCATIONS.with(|t| t.get());
        TRACK_ALLOCATIONS.with(|t| t.set(false));

        Self { previous_state }
    }
}

impl Drop for ScopedToggle {
    fn drop(&mut self) {
        TRACK_ALLOCATIONS.with(|t| t.set(self.previous_state));
    }
}

unsafe impl GlobalAlloc for TrackingAllocator {
    /// Allocates memory and updates the tracking counters.
    unsafe fn alloc(&self, layout: Layout) -> *mut u8 {
        let track = TRACK_ALLOCATIONS.try_with(|t| t.get()).unwrap_or(true);

        if track {
            // Update the allocated bytes and allocation count atomically. No
            // synchronisation is needed for these counters, so
            // `Ordering::Relaxed` can be used.
            ALLOCATED_BYTES.fetch_add(layout.size(), Ordering::Relaxed);
            ALLOCATION_COUNT.fetch_add(1, Ordering::Relaxed);
        }

        // Use the system allocator to perform the actual allocation.
        unsafe { System.alloc(layout) }
    }

    /// Deallocates memory and updates the tracking counter.
    unsafe fn dealloc(&self, ptr: *mut u8, layout: Layout) {
        let track = TRACK_ALLOCATIONS.try_with(|t| t.get()).unwrap_or(true);

        if track {
            // Update the freed bytes counter atomically.
            FREED_BYTES.fetch_add(layout.size(), Ordering::Relaxed);
        }

        // Use the system allocator to perform the actual deallocation.
        unsafe { System.dealloc(ptr, layout) }
    }
}
